log4j-core/src/main/java/org/apache/logging/log4j/core/appender/AsyncAppenderEventForwarder.java

```diff
@@ -0,0 +1,147 @@
+
+    private final BlockingQueue<LogEvent> queue;
+
+    private volatile boolean stopped;
+
+    AsyncAppenderEventForwarder(
+            final String name,
+            final AppenderControl errorAppender,
+            final List<AppenderControl> appenders,
+            final BlockingQueue<LogEvent> queue) {
+        super("AsyncAppenderEventForwarder-" + THREAD_COUNTER.incrementAndGet() + "-" + name);
+        this.errorAppender = errorAppender;
+        this.appenders = appenders;
+        this.queue = queue;
+        this.stopped = false;
+    }
+
+    @Override
+    public void run() {
+        LOGGER.trace("{} has started.", getName());
+        forwardAll();
+        forwardRemaining();
+    }
+
+    private void forwardAll() {
+        while (!stopped) {
+            LogEvent event;
+            try {
+                event = queue.take();
+            } catch (final InterruptedException ignored) {
+                // Restore the interrupted flag cleared when the exception is caught.
+                interrupt();
+                break;
+            }
+            event.setEndOfBatch(queue.isEmpty());
+            forwardOne(event);
+        }
+        LOGGER.trace("{} has stopped.", getName());
+    }
+
+    private void forwardRemaining() {
+        int eventCount = 0;
+        while (true) {
+            // Note the non-blocking Queue#poll() method!
+            final LogEvent event = queue.poll();
+            if (event == null) {
+                break;
+            }
+            event.setEndOfBatch(queue.isEmpty());
+            forwardOne(event);
+            eventCount++;
+        }
+        LOGGER.trace(
+                "{} has processed the last {} remaining event(s).",
+                getName(), eventCount);
+    }
+
+    void forwardOne(final LogEvent event) {
+
+        // Forward the event to all registered appenders.
+        boolean succeeded = false;
+        // noinspection ForLoopReplaceableByForEach (avoid iterator instantion)
+        for (int appenderIndex = 0; appenderIndex < appenders.size(); appenderIndex++) {
+            final AppenderControl control = appenders.get(appenderIndex);
+            try {
+                control.callAppender(event);
+                succeeded = true;
+            } catch (final Throwable ignored) {
+                // If no appender is successful, the error appender will get it.
+            }
+        }
+
+        // Fallback to the error appender if none has succeeded so far.
+        if (!succeeded && errorAppender != null) {
+            try {
+                errorAppender.callAppender(event);
+            } catch (final Throwable ignored) {
+                // If the error appender also fails, there is nothing further
+                // we can do about it.
+            }
+        }
+
+    }
+
+    void stop(final long timeoutMillis) throws InterruptedException {
+
+        // Mark the completion, if necessary.
+        synchronized (this) {
```

Comment:
I think we can replace this synchronization with an atomic getAndSet (since there are only two possible values) or compareAndSet using AtomicBoolean or an AtomicIntegerFieldUpdater (flat object layout, avoids pointer indirection).

Reply:
Doh! :facepalm: A relic from the refactoring. Fixed by replacing with an `AtomicBoolean`.